pkg/tfbridge/provider.go

```diff
@@ -331,44 +331,62 @@ func (p *Provider) GetSchema(ctx context.Context,
 
 // CheckConfig validates the configuration for this Terraform provider.
 func (p *Provider) CheckConfig(ctx context.Context, req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "CheckConfig is not yet implemented")
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.CheckConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
 
-	// TO_DO - revert this comment!!
-	//urn := resource.URN(req.GetUrn())
-	//label := fmt.Sprintf("%s.CheckConfig(%s)", p.label(), urn)
-	//glog.V(9).Infof("%s executing", label)
-	//
-	//news, validationErrors := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
-	//	Label:        fmt.Sprintf("%s.news", label),
-	//	KeepUnknowns: true,
-	//	SkipNulls:    true,
-	//	RejectAssets: true,
-	//})
-	//if validationErrors != nil {
-	//	return nil, errors.Wrap(validationErrors, "CheckConfig failed because of malformed resource inputs")
-	//}
-	//
-	//config, validationErrors := buildTerraformConfig(p, news)
-	//if validationErrors != nil {
-	//	return nil, errors.Wrap(validationErrors, "could not marshal config state")
-	//}
-	//
-	//if p.info.PreConfigureCallback != nil {
-	//	if validationErrors = p.info.PreConfigureCallback(news, config); validationErrors != nil {
-	//		return nil, validationErrors
-	//	}
-	//}
-	//
-	//// This replicates the flow in the validateProviderConfig func where we check for missingKeys first
-	//missingKeys, validationErrors := validateProviderConfig(ctx, p, config)
-	//if len(missingKeys) > 0 {
-	//	return &pulumirpc.CheckResponse{Inputs: req.GetNews(), Failures: missingKeys}, nil
-	//}
-	//if validationErrors != nil {
-	//	return nil, validationErrors
-	//}
-	//
-	//return &pulumirpc.CheckResponse{Inputs: req.GetNews()}, nil
+	marshalOptions := plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.news", label),
+		KeepUnknowns: true,
+		SkipNulls:    true,
+		RejectAssets: true,
+	}
+
+	news, validationErrors := plugin.UnmarshalProperties(req.GetNews(), marshalOptions)
```

Comment:
I'm worried about this and need to understand what tests to add.

Logically it'd make sense to call PreConfigureCallbacks before validation, and it would make sense to validate in CheckConfig. This is a change though in how we parse the values coming into validation funcs.

There is this distinction between variables and args. CheckConfig is now parsing args because it really has no access to variables. But Configure used to parse variables. The parsing functions are different, filtering on prefix is different, and I worry nested object encodings are different. Something might break here.

```
{
  "method": "/pulumirpc.ResourceProvider/Configure",
  "request": {
    "variables": {
      "gcp:config:project": "pulumi-development",
      "gcp:config:redisCustomEndpoint": "http://fromprogram.local/"
    },
    "args": {
      "project": "pulumi-development",
      "redisCustomEndpoint": "http://fromprogram.local/",
      "version": "6.54.0"
    },
    "acceptSecrets": true,
    "acceptResources": true
  },
  "response": {
    "supportsPreview": true
  },
  "metadata": {
    "kind": "resource",
    "mode": "client",
    "name": "gcp"
  }
}
```

Reply:
Added a few tests extracted from end-to-end testing.